### DISCUSSION TITLE
Diffrent node number in mini_batch

### DISCUSSION THREAD

**Discussion by XiaoXiaoYi123** [pyg-team/pytorch_geometric#4179] [Q&A]
Hello, I am fairly new to graph neural networks and I am training a GNN model using self attention and I have a few questions.

The question is my node count and node_num differs in each batch such that in the first batch I have:

`Batch(batch=[1181], edge_attr=[1975, 3], edge_index=[2, 1975], x=[1181, 300])`

in the second batch I have:
`batch=[1134], edge_attr=[1635, 3], edge_index=[2, 1635], x=[1134, 300]`

There were 1181 nodes in batch1, whereas 1134 nodes in batch2

When I tried to calculate self attention between nodes, I encountered the following problem

Here's how self attention works

![image](https://user-images.githubusercontent.com/45426458/156282537-f2c4caa2-e208-454c-b488-665db27f5cd2.png)

the Q, W, K calculate as follows:
![image](https://user-images.githubusercontent.com/45426458/156282696-1029f0d7-6f59-4281-8a17-bf1adf3e4438.png)

the dimension of  wq, wk, wv is 

`self.w_1 = Param(torch.Tensor(self.nodes_num, self.nodes_num))`

So the problem I have is this

- in batch1 , the the dimension of  wq, wk, wv is  `self.w_q = Param(torch.Tensor(1181, 1181))`
-  in batch2 , the the dimension of  wq, wk, wv is  `self.w_q = Param(torch.Tensor(1134, 1134))`

Dimensions vary with the number of nodes, causing w_q to be constantly redefined

Is this equivalent to using only one batch of samples for the model?

If so, how can I solve the problem?

**Comment by rusty1s** [comment]
Note that the parameters do not depend on the number of nodes, but instead on the number of features per node. The transformation are applied in a shared fashion across all nodes, which lets us generalize to any amount of nodes. In your case, `W_q` is a matrix of shape `[in_channels, out_channels]`. Let me know if this clarifies your issues.

**Comment by XiaoXiaoYi123** [comment]
in the first batch I have 'Batch(batch=[1181], edge_attr=[1975, 3], edge_index=[2, 1975], x=[1181, 300])'
in the second batch i have  'batch=[1134], edge_attr=[1635, 3], edge_index=[2, 1635], x=[1134, 300]'
the matrix `wq` i initiate is 'self.w_q = Param(torch.Tensor(self.nodes_num, self.nodes_num))'

The dimension of WQ is consistent with the number of nodes in batch
However, the number of nodes in Mini Batch is not the same, as shown above, BATCH1 = 1181, BATCH2 = 1134
Besides, the X dimension here is [nodes_num, nodes_feature]
The 'wq' size in Batch1 will become [1181][1181]
The 'wq' size in Batch2 will become [1134][1134]

This will cause the matrix to change in size.

Beside what I want to do is to consider the different important of nodes using self-attention

**Reply by rusty1s** [reply]
As I said, the parameters `W_*` do not depend the number of nodes. After multiplication with the initially given node features, `Q`, `K` and `V` will have shape `[num_nodes, out_channels]`. `QK^T` will thus give you the attention matrix of shape `[num_nodes, num_nodes]`. Note that parameters still do not depend on the number of nodes!

**Reply by XiaoXiaoYi123** [reply]
oK,thanks a lot
